contracts/vaults/TreasuryVault/RibbonAutocallVault.sol

```diff
@@ -0,0 +1,444 @@
+            return 0;
+        } else if (_nOptionType == OptionType.DIP) {
+            return _price - _nextStrikePrice;
+        }
+
+        return 0;
+    }
+
+    /**
+     * @dev Returns coupons earned info
+     * @param nCBBreaches the number of observations above coupon barrier
+     * @param lastCBBreach the last observation to breach coupon barrier
+     * @return nCouponsEarned is the number of coupons earned
+     * @return earnedAmt is the autocall buyer earned amount
+     * @return returnAmt is the amount to return to autocallSeller
+     */
+    function _couponsEarned(uint256 nCBBreaches, uint256 lastCBBreach)
+        internal
+        view
+        returns (
+            uint256 nCouponsEarned,
+            uint256 earnedAmt,
+            uint256 returnAmt
+        )
+    {
+        uint256 totalPremium =
+            IERC20(vaultParams.asset).balanceOf(address(this)) -
+                vaultState.totalPending;
+
+        /**
+         * FIXED: coupon barrier is 0, so nCBBreaches will always equal
+         *        total number of observations
+         * PHOENIX: only get the coupon for observations where the spot was above coupon barrier
+         * PHOENIX_MEMORY: the last coupon barrier breach observation will get us the total coupons
+         *                 earned
+         * VANILLA: coupon barrier = autocall barrier so the last coupon barrier breach observation
+         *                 being non-zero means autocall barrier has been hit and we get all previous
+         *                 coupons
+         */
+        bool hasMemory =
+            (couponState.couponType == CouponType.PHOENIX_MEMORY ||
+                couponState.couponType == CouponType.VANILLA)
+                ? true
+                : false;
+        nCouponsEarned = hasMemory ? lastCBBreach : nCBBreaches;
+        earnedAmt = (totalPremium * nCouponsEarned) / nTotalObs;
+        returnAmt = totalPremium - earnedAmt;
+    }
+
+    /**
+     * @dev Gets autocall state
+     * @param _expiry is the expiry of the current option
+     * @return autocallTS the timestamp of first autocallable observation period, otherwise returns _expiry
+     * @return nCBBreaches the number of observations above coupon barrier
+     * @return lastCBBreach the last observation to breach coupon barrier
+     */
+    function _autocallState(uint256 _expiry)
+        internal
+        view
+        returns (
+            uint256 autocallTS,
+            uint256 nCBBreaches,
+            uint256 lastCBBreach
+        )
+    {
+        uint256 startTS = _expiry - (nTotalObs - 1) * obsFreq;
+        (, uint256 lastTS) = _lastObservation(_expiry);
+        address underlying = vaultParams.underlying;
+
+        autocallTS = _expiry;
+
+        // For every previous observation timestamp
+        for (uint256 ts = startTS; ts <= lastTS; ts += obsFreq) {
+            uint256 obsPrice = ORACLE.getExpiryPrice(underlying, ts);
```

Comment:
Wondering whether we should require `obsPrice` to be >0 for the case in which expiry time for a given timestamp has not been defined?